lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs

```diff
@@ -254,26 +279,140 @@ instance ((PersistPublicKey (key 'AccountK)), Typeable chain)
     build (AddressPool acct (AddressPoolGap g) _) = mempty
         <> ccF <> " " <> acctF <> " (gap=" <> build g <> ")\n"
       where
-        ccF = build $ toText $ accountingStyle @chain
+        ccF = build $ toText $ role @chain
+        xpubF = hexF $ serializeXPub acct
+        acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
+
+data VerificationKeyPool (key :: Depth -> * -> *) = VerificationKeyPool {
+      verPoolAccountPubKey
+        :: !(key 'AccountK XPub)
+    -- ^ Corresponding key for the pool (a pool is tied to only one account)
+    , verPoolGap
+        :: !AddressPoolGap
+    -- ^ The actual gap for the pool. This can't change for a given pool.
+    , verPoolIndexedKeys
+        :: !(Map KeyHash (Index 'Soft 'ScriptK, AddressState))
+    , verPoolKnownScripts :: !(Map ScriptHash [Index 'Soft 'ScriptK])
+        -- ^ Known script hashes that contain our verification key hashes
+        -- represented here by corresponding indices
+    } deriving (Generic)
+
+deriving instance (Show (key 'AccountK XPub), Show (key 'ScriptK XPub))
+    => Show (VerificationKeyPool key)
+
+deriving instance (Eq (key 'AccountK XPub), Eq (key 'ScriptK XPub))
+    => Eq (VerificationKeyPool key)
+
+instance (NFData (key 'AccountK XPub), NFData (key 'ScriptK XPub))
+    => NFData (VerificationKeyPool key)
+
+instance ((PersistPublicKey (key 'AccountK)))
+    => Buildable (VerificationKeyPool key) where
+    build (VerificationKeyPool acct (AddressPoolGap g) _ _) = mempty
+       <> ccF <> " " <> acctF <> " (gap=" <> build g <> ")\n"
+      where
+        ccF = build $ toText MultisigScript
         xpubF = hexF $ serializeXPub acct
         acctF = prefixF 8 xpubF <> "..." <> suffixF 8 xpubF
 
--- | Bring a 'AccountingStyle' type back to the term-level. This requires a type
+deriving instance Ord KeyHash
+
+deriveKeyHash
+    :: (WalletKey k, SoftDerivation k)
+    => k 'AccountK XPub
+    -> Index 'Soft 'ScriptK
+    -> KeyHash
+deriveKeyHash accXPub' =
+    toVerKeyHash . deriveAddressPublicKey accXPub' MultisigScript . coerce
```

Comment:
a) Shouldn't this be part of the AddressDerivation module instead?
b) Seems like this function is doing 2 things at once: deriving & hashing. It'd be better to keep them separate and simply compose them in place.
c) I think it should be the caller's responsibility to use coerce if deemed safe, but that function should use only ScriptK` in its signature. 

Note that I don't think there's a need for such a function to be frank. Instead, it should suffice to make the `role` a type parameter on the `SoftDerivation` class definition (instead of being fixed to 'AddressK'

Reply:
added `deriveVerificationKey` and `calculateVerificationKeyHash` in AddressDerivation. In order to avoid Coercible needed to leave `liftRawKey`. Tried to make `role` type parameter in `SoftDerivation` but it produced considerable number of ripple effects. so stick with straightforward way